### ISSUE TITLE
Is there anyway to pass handler informations into middleware to let it decide between different behaviours?

### ISSUE THREAD

**Issue by GoodManWEN** [tiangolo/fastapi#2810]
### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI

app = FastAPI()
@app.middleware("http")
async def auth_detect(request: Request, call_next):
    if call_next.require_auth
        if request.headers.authname in ('admin' , 'guest'):
            response = await call_next(request)
        else:
            response = HTTPException(status_code=404, detail="Item not found")
    else:
        response = await call_next(request)
    return response

@require_no_auth()  # define middleware behaviour on defination.
@app.get("/")
def root():
    return {"Hello": "World"}

@require_auth()  
@app.get("/secret")
def secret():
    return {"Some": "Secret"}

```

### Description

Hi, I'm currently wondering if there's any way to control different behaviour between different routes in the same middleware. For example if I'd like to return a status code of if the requested page is auth required , and pass directly if there's no authentication needed , above-mentioned in my pseudo code, didn't find much info in docs , need help , thanks.

**Comment by frankie567** [comment]
I think a more idiomatic way to do this in FastAPI is to use dependencies instead of middlewares:

```py
from fastapi import Depends, FastAPI, Header, HTTPException

app = FastAPI()

async def require_auth(authname: str = Header(...)):
    if authname not in ("admin", "guest"):
        raise HTTPException(status_code=404, detail="Item not found")


@app.get("/")
def root():
    return {"Hello": "World"}


@app.get("/secret", dependencies=[Depends(require_auth)])
def secret():
    return {"Some": "Secret"}
```

You specify this dependency on endpoints that require authentication. This dependency checks if the header is set and correct, otherwise it raises an `HTTPException`.

Running example: 
* REPL: https://repl.it/@frankie567/fastapi-issue-2810
* API Endpoint: https://fastapi-issue-2810--frankie567.repl.co/secret

**Comment by GoodManWEN** [comment]
@frankie567  Thanks for your kind reply! Due to the experience of using other frameworks I naturally thought of using middleware to solve this problem, But as you said, using dependencies in fastapi could really be a better solution ,thanks.

**Comment by GoodManWEN** [comment]
But I still have a problem with modularity . As I understand how it is used ,  I'd like to extract a separate file which contains all the dependencies (just like middle wares), meanwhile I got another file for all the webviews(contains the main business codes) on the other hands , for example I got a `handlers.py` like this
```
# handlers.py
from fastapi.responses import UJSONResponse , HTMLResponse
from fastapi import Request
from .dependencies import auth_checker

...

@app.get("/secret", dependencies=[Depends(auth_checker)])
async def test_api(request: Request):
    return UJSONResponse({'hello':'world'})

@app.get('/')
async def index():
    return HTMLResponse('<h1>Index Page</h1>')
```

However depending on how the dependency is used , there are times when I need it to return an handler but not some bare response of status_code 404 or something , for example I need it return `index()` but not 404 If authentication fails.

In that case, when defining the dependency of `auth_checker` it should import `index()` from `.handlers.py` and handlers.py should also import `auch_checker` from `.dependencies.py` , thus gives rise to a circular import problem. And due to fastapi's special implementation of inputs to the handler functions , this problem cannot be solved by some other intermediate transmission workaround.

What a reasonable modular approach should look like  when using dependencies? I could write them all in the same file, but then the file would be too long.

**Comment by frankie567** [comment]
Well, IMO, this is a bit weird to return the response of another handler when something goes wrong on one route (RESTful API should return clear status codes about what happened, and a single resource should live on a single endpoint).

That said, you could very well define your `index` response in a dependency and return it in your router logic:

```py
from typing import Optional

from fastapi import Depends, FastAPI, Header
from fastapi.responses import HTMLResponse, UJSONResponse

app = FastAPI()


async def get_index_response() -> HTMLResponse:
  return HTMLResponse('<h1>Index Page</h1>')


async def require_auth(
    authname: Optional[str] = Header(None),
    index_response: HTMLResponse = Depends(get_index_response)
) -> Optional[HTMLResponse]:
    if not authname or authname not in ("admin", "guest"):
        return index_response


@app.get("/secret")
async def test_api(require_auth_response: Optional[HTMLResponse] = Depends(require_auth)):
    if require_auth_response:
      return require_auth_response
    return UJSONResponse({'hello':'world'})


@app.get('/')
async def index(index_response: HTMLResponse = Depends(get_index_response)):
    return index_response
```

But again, this is quite a weird way to design a REST API.

**Comment by GoodManWEN** [comment]
Yes indeed , can feel that fastapi is a engine specially designed for "APIs". What I'm doing is trying to use it to serve a web application , maybe a little inappropriate use but there are some features , including asynchronous , of fastapi that really appeal to me.

**Comment by tiangolo** [comment]
Thanks for the help here @frankie567 ! 👏  🙇 

Thanks for reporting back and closing the issue @GoodManWEN  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
